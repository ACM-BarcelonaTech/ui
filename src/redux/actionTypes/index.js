const CELLS = 'CELLS';
const LOAD_CELLS = `${CELLS}.LOAD`;

const GENE_LIST = 'GENE_LIST';
const LOAD_GENE_LIST = `${GENE_LIST}.LOAD`;
const UPDATE_GENE_LIST = `${GENE_LIST}.UPDATE`;

const DIFF_EXPR = 'DIFF_EXPR';
const LOAD_DIFF_EXPR = `${DIFF_EXPR}.LOAD`;
const UPDATE_DIFF_EXPR = `${DIFF_EXPR}.UPDATE`;

const SELECTED_GENES = 'SELECTED_GENES';

const BUILD_HEATMAP_SPEC = 'BUILD_HEATMAP_SPEC';
const UPDATE_GENE_EXPRESSION = 'UPDATE_GENE_EXPRESSION';
const UPDATE_HEATMAP_SPEC = 'UPDATE_HEATMAP_SPEC';

const UPDATE_CELL_INFO = 'UPDATE_CELL_INFO';

const SET_FOCUSED_GENE = 'SET_FOCUSED_GENE';

export {
  LOAD_CELLS,
  LOAD_GENE_LIST, UPDATE_GENE_LIST, SELECTED_GENES, BUILD_HEATMAP_SPEC,
  UPDATE_GENE_EXPRESSION, UPDATE_HEATMAP_SPEC, LOAD_DIFF_EXPR, UPDATE_DIFF_EXPR,
  UPDATE_CELL_INFO, SET_FOCUSED_GENE,
};
