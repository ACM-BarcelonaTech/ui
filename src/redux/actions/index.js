/* eslint-disable no-param-reassign */
import { v4 as uuidv4 } from 'uuid';
import {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, PUSH_CELL_SETS, CREATE_CLUSTER, CELL_SETS_COLOR,
  UPDATE_GENE_LIST, LOAD_GENE_LIST, SELECTED_GENES, UPDATE_GENE_EXPRESSION,
  LOAD_CELLS, BUILD_HEATMAP_SPEC, UPDATE_HEATMAP_SPEC, LOAD_DIFF_EXPR, UPDATE_DIFF_EXPR,
} from './actionType';
import connectionPromise from '../../utils/socketConnection';

const loadCellSets = (experimentId) => (dispatch, getState) => {
  if (getState().cellSets.data) {
    return null;
  }

  return fetch(`${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: LOAD_CELL_SETS,
      experimentId,
      data: json.cellSets,
    }),
  ).catch((e) => console.error('Error when trying to get cell sets data: ', e));
};

const updateCellSets = (experimentId, newState) => (dispatch, getState) => {
  if (getState().cellSets.data === newState) {
    return Promise.resolve();
  }

  dispatch({
    type: UPDATE_CELL_SETS,
    experimentId,
    data: newState,
  });

  return dispatch(pushCellSets(experimentId));
};

const pushCellSets = (experimentId) => (dispatch, getState) => {
  // If no loaded data exists for our cell sets and some event would
  // trigger a push, we do not want to overwrite our cell sets with
  // empty data.
  if (!getState()?.cellSets?.data) {
    return null;
  }

  return fetch(
    `${process.env.REACT_APP_API_URL}/v1/experiments/${experimentId}/cellSets`,
    {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(
        getState().cellSets.data,
        (k, v) => ((k === 'title') ? undefined : v),
      ),
    },
  ).then(
    (response) => response.json(),
  ).then(
    (json) => dispatch({
      type: PUSH_CELL_SETS,
      experimentId,
      data: json,
    }),
  ).catch((e) => console.error('Error when trying to push cell sets to API: ', e));
};

const createCluster = (experimentId, cellSetInfo, clusterName, color) => (dispatch) => {
  const clusterKey = uuidv4();
  const newCluster = {
    key: clusterKey,
    name: clusterName,
    color,
    cellIds: Array.from(cellSetInfo),
  };

  dispatch({
    type: CREATE_CLUSTER,
    experimentId,
    data: newCluster,
  });

  return dispatch(pushCellSets(experimentId));
};

const cellSetsColor = (colorData) => (dispatch, getState) => {
  if (getState().cells.data) {
    return dispatch({
      type: CELL_SETS_COLOR,
      data: colorData,
    });
  }
  return null;
};

const loadCells = (experimentId, embeddingType) => (dispatch, getState) => {
  if (getState().cells.data) {
    return null;
  }
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();

    const body = {
      name: 'GetEmbedding',
      type: embeddingType,
    };

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const embedding = JSON.parse(res.results[0].body);
      return dispatch({
        type: LOAD_CELLS,
        experimentId,
        data: embedding,
      });
    });
  });
};

const updateGeneList = (experimentId, tableState) => (dispatch, getState) => {
  if (getState().geneList?.tableState === tableState) {
    return null;
  }
  dispatch({
    type: LOAD_GENE_LIST,
  });
  return connectionPromise().then((io) => {
    const requestUuid = uuidv4();
    const { geneList } = getState();

    const orderBy = tableState?.sorter.field
      || geneList.tableState?.sorter.field
      || 'dispersions';

    const orderDirection = tableState?.sorter.order
      || geneList.tableState?.sorter.order
      || 'ascend';

    const currentPage = tableState?.pagination.current
      || geneList.tableState?.pagination.current
      || 1;

    const currentPageSize = tableState?.pagination.pageSize
      || geneList.tableState?.pagination.pageSize
      || 1;

    const body = {
      name: 'ListGenes',
      selectFields: ['gene_names', 'dispersions'],
      orderBy,
      orderDirection: (orderDirection === 'ascend') ? 'ASC' : 'DESC',
      offset: ((currentPage - 1) * currentPageSize),
      limit: currentPageSize,
    };

    if (tableState.geneNamesFilter) {
      body.geneNamesFilter = tableState.geneNamesFilter;
    }

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);
    io.on(`WorkResponse-${requestUuid}`, (res) => {
      const data = JSON.parse(res.results[0].body);

      const { total, rows } = data;

      rows.map((row) => {
        row.key = row.gene_names;
        return row;
      });

      if (tableState && tableState.pagination) {
        tableState.pagination.total = total;
      }
      return dispatch({
        type: UPDATE_GENE_LIST,
        experimentId,
        data: {
          rows,
          tableState: tableState || geneList.tableState,
        },
      });
    });
  });
};

const loadDiffExpr = (
  experimentId, comparisonType, firstSelectedCluster, secondSelectedCluster,
) => (dispatch) => {
  dispatch({
    type: LOAD_DIFF_EXPR,
  });
  return connectionPromise().then((io) => {
    const ComparisonTypes = {
      One: 'Versus Rest',
      Two: 'Across Sets',
    };

    const requestUuid = uuidv4();

    const body = {
      name: 'DifferentialExpression',
      maxNum: 100,
      cellSet: firstSelectedCluster.key,
    };
    if (comparisonType === ComparisonTypes.One) {
      body.compareWith = 'rest';
    } else {
      body.compareWith = secondSelectedCluster.key;
    }

    const request = {
      uuid: requestUuid,
      socketId: io.id,
      experimentId,
      timeout: '2021-01-01T00:00:00Z',
      body,
    };

    io.emit('WorkRequest', request);

    io.on(`WorkResponse-${requestUuid}`, (res) => {
      let data = {};

      try {
        data = JSON.parse(res.results[0].body);
      } catch (error) {
        console.error(error);
        data = { rows: [] };
      }

      const { rows } = data;
      const total = rows.length;

      rows.map((row) => {
        row.key = row.gene_names;
        return row;
      });

      return dispatch({
        type: UPDATE_DIFF_EXPR,
        experimentId,
        data: {
          allData: rows,
          total,
        },
      });
    });
  });
};

const updateSelectedGenes = (genes, selected) => (dispatch, getState) => {
  const { selectedGenes, geneExperessionData } = getState();
  selectedGenes.geneList = selectedGenes.geneList || {};
  let newGenesAdded = false;
  genes.forEach((gene) => {
    if (selected) {
      if (!selectedGenes.geneList[gene]) {
        // This will present a dict of newly selected genes from gene table
        // { <Gene Name>: <Is the gene experession resolved?> }
        selectedGenes.geneList[gene] = false;
        newGenesAdded = true;
      }
    } else {
      delete selectedGenes.geneList[gene];
      const foundGene = geneExperessionData.data.findIndex((g) => g.geneName === gene);
      geneExperessionData.data.splice(foundGene, 1);
    }
  });
  dispatch({
    type: SELECTED_GENES,
    data: { newGenesAdded },
  });
  dispatch({
    type: UPDATE_HEATMAP_SPEC,
    data: {
      genes: geneExperessionData.data,
      rendering: true,
    },
  });
  setTimeout(() => {
    // Running this action with a delay to allow rerendering of heatmap
    dispatch({
      type: UPDATE_HEATMAP_SPEC,
      data: { rendering: false },
    });
  }, 50);
};

const loadGeneExpression = (experimentId) => (dispatch, getState) => {
  const { geneExperessionData } = getState();
  if (geneExperessionData?.isLoading) {
    return null;
  }
  const { newGenesAdded, geneList } = getState().selectedGenes;
  const newGeneBatch = [];
  if (newGenesAdded) {
    // eslint-disable-next-line no-restricted-syntax
    for (const [key, value] of Object.entries(geneList)) {
      if (!value) {
        newGeneBatch.push(key);
        geneList[key] = true;
      }
    }
  }
  dispatch({
    type: SELECTED_GENES,
    data: {
      newGenesAdded: false,
    },
  });
  if (newGeneBatch.length) {
    dispatch({
      type: UPDATE_GENE_EXPRESSION,
      data: {
        isLoading: true,
      },
    });
    return connectionPromise().then((io) => {
      const requestUuid = uuidv4();

      const body = {
        name: 'GeneExpression',
        cellSets: 'all',
        genes: newGeneBatch,
      };

      const request = {
        uuid: requestUuid,
        socketId: io.id,
        experimentId,
        timeout: '2021-01-01T00:00:00Z',
        body,
      };

      io.emit('WorkRequest', request);

      io.on(`WorkResponse-${requestUuid}`, (res) => {
        const heatMapData = JSON.parse(res.results[0].body);
        const { data } = getState().geneExperessionData;
        if (data) {
          Array.prototype.push.apply(heatMapData.data, data);
        }
        dispatch({
          type: UPDATE_GENE_EXPRESSION,
          data: {
            heatMapData,
            isLoading: false,
          },
        });
        return dispatch({
          type: BUILD_HEATMAP_SPEC,
          data: {
            geneExperessionData: heatMapData,
          },
        });
      });
    });
  }
};

export {
  loadCellSets,
  updateCellSets,
  pushCellSets,
  createCluster,
  loadCells,
  cellSetsColor,
  updateGeneList,
  updateSelectedGenes,
  loadGeneExpression,
  loadDiffExpr,
};
