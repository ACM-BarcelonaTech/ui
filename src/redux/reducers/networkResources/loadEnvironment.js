/* eslint-disable no-param-reassign */
import produce from 'immer';

const loadEnvironment = produce((draft, action) => {
  const { environment } = action.payload;
  draft.environment = environment;
});

export default loadEnvironment;
