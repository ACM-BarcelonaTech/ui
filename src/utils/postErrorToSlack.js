import Auth from '@aws-amplify/auth';
import { getLoggerBotToken } from 'utils/crypt';

const extractExperimentId = (url) => {
  const match = url.match(/experiments\/([^/]+)/i);
  return match ? match[1] : null;
};

const takeNumLines = (string, numLines) => string.split('\n').slice(0, numLines).join('\n');

const buildErrorMessage = (error, componentStack, reduxDump, context) => {
  const {
    user, timestamp, experimentId, url,
  } = context;

  return (`
    Uncaught UI Error - Exp ID ${experimentId} - ${timestamp}

    === DETAILS ===
    Name: ${user.attributes.name}
    Email: ${user.attributes.email}
    UserID: ${user.username}
    ExperimentID: ${experimentId}
    URL: ${url}
    Timestamp: ${timestamp}

    ===== ERROR =====
    ${error.toString()}
    ${takeNumLines(componentStack, 11)}

    ===== REDUX STATE =====
    ${JSON.stringify(reduxDump, null, 2)}`
  );
};

const postError = async (errorLog, context) => {
  const {
    user, timestamp, experimentId, url,
  } = context;

  const message = `
  \u26A0  Uncaught UI Error - ExpID ${experimentId} - ${timestamp}
  URL: ${url}

  User: ${user.attributes.name}
  Email: ${user.attributes.email}
  User ID: ${user.username}
  Experiment ID: ${experimentId}`;

  const formData = new FormData();

  const formFields = {
    token: getLoggerBotToken(),
    title: `UI Error at ${new Date().toISOString()}`,
    initial_comment: message,
    channels: 'error-logs',
    filename: `UI-error-${experimentId}-${timestamp}.log`,
    content: errorLog,
  };

  Object.entries(formFields).forEach(([fieldName, value]) => {
    formData.append(fieldName, value);
  });

  try {
    const res = await fetch('https://slack.com/api/files.upload', {
      method: 'POST',
      body: formData,
    });

    if (!res.ok) {
      console.error('Failed sending error to slack');
    }
  } catch (err) {
    console.error(err);
  }
};

const postErrorToSlack = async (error, info, reduxDump) => {
  const user = await Auth.currentAuthenticatedUser();

  const timestamp = new Date().toISOString();
  const url = window.location.href;
  const experimentId = extractExperimentId(url);
  const { componentStack } = info;

  const context = {
    user,
    timestamp,
    experimentId,
    url,
  };

  const errorLog = buildErrorMessage(error, componentStack, reduxDump, context);
  await postError(errorLog, context);
};

export default postErrorToSlack;
