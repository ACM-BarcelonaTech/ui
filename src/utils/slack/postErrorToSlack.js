import Auth from '@aws-amplify/auth';
import postToSlack from 'utils/slack/postToSlack';

const getPageContext = () => [
  {
    type: 'mrkdwn',
    text: '*URL posted from:*',
  },
  {
    type: 'mrkdwn',
    text: window.location.href,
  },
];

const getUserContext = async () => {
  let user = null;

  try {
    user = await Auth.currentAuthenticatedUser();
  } catch (e) {
    return [
      {
        type: 'mrkdwn',
        text: 'User not authenticated',
      },
    ];
  }

  return [
    {
      type: 'mrkdwn',
      text: '*User email:*',
    },
    {
      type: 'mrkdwn',
      text: user.attributes.email,
    },

    {
      type: 'mrkdwn',
      text: '*User name:*',
    },
    {
      type: 'plain_text',
      text: user.attributes.name,
    },

    {
      type: 'mrkdwn',
      text: '*User UUID:*',
    },
    {
      type: 'plain_text',
      text: user.username,
    },
  ];
};

const buildErrorMessage = (error, info) => [
  {
    type: 'mrkdwn',
    text: '*Error:*',
  },
  {
    type: 'mrkdwn',
    text: error.message,
  },
  {
    type: 'mrkdwn',
    text: `\`\`\`${error.stack}\`\`\``,
  },
  {
    type: 'mrkdwn',
    text: '*the above error occured in: *',
  },
  {
    type: 'mrkdwn',
    text: `\`\`\`${info.ComponentStack}\`\`\``,
  },
];

const postErrorToSlack = async (error, info) => {
  const pageContext = getPageContext();
  const userContext = await getUserContext();
  const errorMessage = buildErrorMessage(error, info);

  const messageData = {
    blocks: [
      {
        type: 'section',
        text: {
          type: 'plain_text',
          text: errorMessage,
        },
      },
      {
        type: 'context',
        elements: [
          ...pageContext,
          ...userContext,
        ],
      },
    ],
  };

  try {
    postToSlack(messageData);
  } catch (err) {
    console.error(err);
  }
};

export default postErrorToSlack;
