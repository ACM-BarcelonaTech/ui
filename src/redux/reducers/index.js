/* eslint-disable no-param-reassign */
import { combineReducers } from 'redux';
import {
  LOAD_CELL_SETS, UPDATE_CELL_SETS, CREATE_CLUSTER, LOAD_CELLS,
  CELL_SETS_COLOR, UPDATE_GENE_LIST, LOAD_GENE_LIST,
  LOAD_DIFF_EXPR, UPDATE_DIFF_EXPR,
} from '../actions/actionType';


const cellSetsReducer = (state = {}, action) => {
  switch (action.type) {
    case LOAD_CELL_SETS:
      state.data = action.data;
      return state;
    case UPDATE_CELL_SETS:
      state.data = action.data;
      return state;
    case CREATE_CLUSTER:
      // for now, if cell set tool is not opened yet, we do nothing on create cell set action
      // in the future, we will need to handle that case too.
      if (!state.data) {
        // Find scratchpad at top level and add the new cluster.
        // The assignment is necessary because otherwise `useSelector`
        // won't recognize the changed state and the cell set tool won't rerender.

        state.data = state.data.map((topCategory) => {
          if (topCategory.key === 'scratchpad') {
            if (!topCategory.children) {
              topCategory.children = [];
            }

            topCategory.children.push(action.data);
          }

          return topCategory;
        });
      }
      return state;
    default:
      return state;
  }
};

const cellSetsColorReducer = (state = {}, action) => {
  switch (action.type) {
    case CELL_SETS_COLOR:
      state.data = action.data;
      return state;
    default:
      return state;
  }
};

const cellsReducer = (state = {}, action) => {
  switch (action.type) {
    case LOAD_CELLS:
      state.data = action.data;
      return state;
    default:
      return state;
  }
};

const geneListReducer = (state = {}, action) => {
  switch (action.type) {
    case LOAD_GENE_LIST:
      return {
        ...state,
        loading: true,
      };
    case UPDATE_GENE_LIST:
      return {
        ...state,
        ...action.data,
        loading: false,
      };
    default:
      return state;
  }
};

const diffExprReducer = (state = {}, action) => {
  switch (action.type) {
    case LOAD_DIFF_EXPR:
      return {
        ...state,
        loading: true,
      };
    case UPDATE_DIFF_EXPR:
      return {
        ...state,
        ...action.data,
        loading: false,
      };
    default:
      return state;
  }
};

export default combineReducers({
  cellSets: cellSetsReducer,
  cellSetsColor: cellSetsColorReducer,
  cells: cellsReducer,
  geneList: geneListReducer,
  diffExpr: diffExprReducer,
});
