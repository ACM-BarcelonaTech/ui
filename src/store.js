import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';
import createRootReducer from './reducers';

export default createStore(createRootReducer, applyMiddleware(thunk));
