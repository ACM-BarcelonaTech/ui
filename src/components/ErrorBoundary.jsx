/* eslint-disable react/destructuring-assignment */
import React from 'react';
import PropTypes from 'prop-types';
import postErrorToSlack from 'utils/postErrorToSlack';
import { connect } from 'react-redux';
import Error from 'pages/_error';

// Implementation of https://reactjs.org/docs/error-boundaries.html
// Using React class because it's not yet supported for functional components
// according to https://stackoverflow.com/a/68075800/1940886
class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError() {
    // Update state so the next render will show the fallback UI.
    return { hasError: true };
  }

  componentDidCatch(error, errorInfo) {
    // Act on the error inside this function
    if (this.props.environment !== 'production') return;
    const { reduxDump } = this.props;
    postErrorToSlack(error, errorInfo, reduxDump);
  }

  render() {
    if (this.state.hasError) return <Error />;

    return this.props.children;
  }
}

function mapStateToProps(state) {
  return {
    environment: state.networkResources.environment,
    reduxDump: state,
  };
}

ErrorBoundary.propTypes = {
  reduxDump: PropTypes.object,
  environment: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
};

ErrorBoundary.defaultProps = {
  reduxDump: {},
};

export default connect(mapStateToProps)(ErrorBoundary);
