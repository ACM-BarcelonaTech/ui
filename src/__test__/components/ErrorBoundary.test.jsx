import React, { useState } from 'react';
import { screen, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';
import { makeStore } from 'redux/store';
import postErrorToSlack from 'utils/postErrorToSlack';

import ErrorBoundary from 'components/ErrorBoundary';
import loadEnvironment from 'redux/actions/networkResources/loadEnvironment';

const content = 'Some content';

// ErrorBoundary renders the `_error.jsx` component if an error occurs.
// This is a string contained in the `_error.jsx component to test if the component is rendered.
const errorText = /Sorry, something went wrong on our end/i;

jest.mock('utils/postErrorToSlack');

delete window.location;
window.location = { reload: jest.fn() };

const throwError = () => {
  screen.getByText('Throw').click();
};

const TestComponent = () => {
  const [hasError, setHasError] = useState(false);

  if (hasError) throw new Error('Error');

  return (
    <>
      <button type='button' onClick={() => setHasError(true)}>Throw</button>
      <div>{content}</div>
    </>
  );
};

const renderErrorBoundary = (store, hasError = true) => {
  render(
    <Provider store={store}>
      <ErrorBoundary>
        <TestComponent hasError={hasError} />
      </ErrorBoundary>
    </Provider>,
  );
};

let storeState = null;

describe('ErrorBoundary', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    storeState = makeStore();
    storeState.dispatch(loadEnvironment('production'));
  });

  it('Does not render if there is no error', () => {
    renderErrorBoundary(storeState, false);

    expect(screen.getByText(content)).toBeInTheDocument();
    expect(screen.queryByText(errorText)).toBeNull();
  });

  it('Still render the UI if there is an error', async () => {
    renderErrorBoundary(storeState);

    throwError();

    await waitFor(() => {
      expect(postErrorToSlack).toHaveBeenCalledTimes(1);
    });

    expect(screen.getByText(content)).toBeInTheDocument();
    expect(screen.queryByText(errorText)).toBeNull();
  });

  it('Should post error if environment is production', async () => {
    renderErrorBoundary(storeState);

    throwError();

    await waitFor(() => {
      expect(postErrorToSlack).toHaveBeenCalledTimes(1);
      const payload = postErrorToSlack.mock.calls[0];
      expect(payload).toMatchSnapshot();
    });
  });

  it('Should not post error if environment is not production', async () => {
    storeState.dispatch(loadEnvironment('staging'));
    renderErrorBoundary(storeState);

    throwError();

    expect(postErrorToSlack).not.toHaveBeenCalled();
  });
});
