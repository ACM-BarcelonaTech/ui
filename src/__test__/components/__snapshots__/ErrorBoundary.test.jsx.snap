// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary Should post error if environment is production 1`] = `
Array [
  [Error: Error],
  Object {
    "componentStack": "
    in TestComponent (at ErrorBoundary.test.jsx:44)
    in ErrorBoundary (created by Connect(ErrorBoundary))
    in Connect(ErrorBoundary) (at ErrorBoundary.test.jsx:43)
    in Provider (at ErrorBoundary.test.jsx:42)",
  },
  Object {
    "backendStatus": Object {},
    "cellInfo": Object {
      "focus": Object {
        "key": "louvain",
        "store": "cellSets",
      },
      "groupedTrack": "louvain",
      "selectedTracks": Array [
        "louvain",
      ],
    },
    "cellMeta": Object {
      "doubletScores": Object {
        "data": Array [],
        "error": false,
        "loading": true,
      },
      "mitochondrialContent": Object {
        "data": Array [],
        "error": false,
        "loading": true,
      },
    },
    "cellSets": Object {
      "error": false,
      "hidden": Set {},
      "hierarchy": Array [],
      "loading": true,
      "properties": Object {},
      "selected": Object {},
      "updatingClustering": false,
    },
    "componentConfig": Object {},
    "differentialExpression": Object {
      "comparison": Object {
        "advancedFilters": Array [],
        "group": Object {
          "between": Object {
            "basis": null,
            "cellSet": null,
            "compareWith": null,
          },
          "within": Object {
            "basis": null,
            "cellSet": null,
            "compareWith": null,
          },
        },
        "ordering": Object {
          "orderBy": "logFC",
          "orderDirection": "DESC",
        },
        "type": "within",
      },
      "properties": Object {
        "comparisonGroup": Object {},
        "comparisonType": null,
        "data": Array [],
        "error": false,
        "loading": false,
        "total": 0,
      },
    },
    "embeddings": Object {},
    "experimentSettings": Object {
      "info": Object {
        "experimentId": null,
        "experimentName": null,
        "sampleIds": Array [],
      },
      "originalProcessing": Object {},
      "processing": Object {
        "meta": Object {
          "changedQCFilters": Set {},
          "completingStepError": false,
          "loading": false,
          "loadingSettingsError": false,
        },
      },
    },
    "experiments": Object {
      "ids": Array [],
      "meta": Object {
        "error": false,
        "loading": false,
        "saving": false,
      },
    },
    "genes": Object {
      "expression": Object {
        "data": Object {},
        "error": false,
        "loading": Array [],
        "views": Object {},
      },
      "focused": undefined,
      "markers": Object {
        "error": false,
        "loading": false,
      },
      "properties": Object {
        "data": Object {},
        "loading": Array [],
        "views": Object {},
      },
      "selected": Array [],
    },
    "layout": Object {
      "panel": "Gene list",
      "windows": Object {
        "direction": "row",
        "first": Object {
          "direction": "column",
          "first": Object {
            "direction": "row",
            "first": "UMAP",
            "second": "Data Management",
            "splitPercentage": 60,
          },
          "second": "Heatmap",
          "splitPercentage": 50,
        },
        "second": "Genes",
        "splitPercentage": 70,
      },
    },
    "networkResources": Object {
      "environment": "production",
    },
    "projects": Object {
      "ids": Array [],
      "meta": Object {
        "activeProjectUuid": null,
        "error": false,
        "loading": true,
        "saving": false,
      },
    },
    "samples": Object {
      "meta": Object {
        "error": false,
        "loading": false,
        "saving": false,
      },
    },
  },
]
`;
